book/src/dev/rfcs/0005-treestate.md

```diff
@@ -0,0 +1,131 @@
+
+**nullifier set**: The set of unique nullifiers revealed by any transactions
+within a block. Nullifiers are enforced to be unique within a valid block chain
+by commiting to previous treestates in Spend descriptions, in order to prevent
+double-spends.
+
+**note commitments**: Pedersen commitment to the values consisting a note. One
+should not be able to construct a note from its commitment.
+
+**note commitment tree**: An incremental Merkle tree of fixed depth used to
+store note commitments that JoinSplit transfers or Spend transfers produce. It
+is used to express the existence of value and the capability to spend it. It is
+not the job of this tree to protect against double-spending, as it is
+append-only: that's what the nullifier set is for.
+
+**note position**: The index of a note commitment at the leafmost layer,
+counting leftmost to rightmost. The [position in the tree is determined by the
+order of transactions in the block](https://zips.z.cash/protocol/canopy.pdf#transactions).
+
+**root**: The layer 0 node of a merkle tree.
+
+**anchor**: A Merkle tree root of a note commitment tree. It uniquely identifies
+a note commitment tree state given the assumed security properties of the Merkle
+tree’s hash function.  Since the nullifier set is always updated together with
+the note commitment tree, this also identifies a particular state of the
+associated nullier set.
+
+**spend descriptions**: A shielded Sapling transfer that spends a note. Includes
+an anchor of some previous block's note commitment tree.
+
+**output descriptions**: A shielded Sapling transfer that creates a
+note. Includes the u-coordinate of the note commitment itself.
+
+**joinsplit**: A shielded transfer that can spend Sprout notes and transparent
+value, and create new Sprout notes and transparent value, in one Groth16 proof
+statement.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As blocks are validated, the note commitments revealed by all the transcations
+within that block are used to construct note commitment trees, with the note
+commitments aligned in their note positions in the bottom layer of the Sprout or
+Sapling note tree from the left-most leaf to the right-most in transaction order
+in the block. So the Sprout note commitments revealed by the first
+JoinSplit<Groth16> in a block would take note position 0 in the Sprout note
+commitment tree, for example. Once all the transactions in a block are parsed
+and the notes for each tree collected in their appropriate positions, the root
+of each tree is computed. While the trees are being built, the respective block
+nullifier sets are updated in memory as note nullifiers are revealed. If the
+rest of the block is validated according to consensus rules, that root is
+committed to its own datastructure via our state service (Sprout anchors,
+Sapling anchors). Sapling block validation includes comparing the specified
```

Comment:
See https://github.com/ZcashFoundation/zebra/pull/902/files#r482674070 for a suggestion of how to store these anchors.

Reply:
Looks good, maybe we should cross-reference here?